File: src/core/cgroup.rs
use std::fs;
use std::os::unix::fs::PermissionsExt;
use std::io;

use crate::config::{CGROUP_ROOT, MEMORY_MAX_MB, MEMORY_SWAP_MAX_MB, CPU_QUOTA_US, CPU_PERIOD_US};

pub const MEMORY_MAX: u64 = MEMORY_MAX_MB * 1024 * 1024;
pub const MEMORY_SWAP_MAX: u64 = MEMORY_SWAP_MAX_MB * 1024 * 1024;

/// Initialize the cgroup root directory (should be called once at daemon startup)
pub fn init_cgroup_root() -> io::Result<()> {
    // Create the main cgroup directory if it doesn't exist
    if !std::path::Path::new(CGROUP_ROOT).exists() {
        fs::create_dir_all(CGROUP_ROOT)?;
        fs::set_permissions(CGROUP_ROOT, fs::Permissions::from_mode(0o755))?;
    }
    
    // Always try to enable memory, cpu, and net_cls controllers (even if dir already exists)
    let subtree_control = format!("{}/cgroup.subtree_control", CGROUP_ROOT);
    if let Err(e) = fs::write(&subtree_control, "+memory +cpu") {
        eprintln!("Warning: Failed to enable cgroup controllers: {}", e);
    } else {
        eprintln!("✓ Cgroup controllers enabled: memory, cpu");
    }
    
    Ok(())
}

/// Setup cgroup for a specific container (called from parent process before fork)
pub fn setup_cgroup_for_container(container_name: &str) -> io::Result<String> {
    init_cgroup_root()?;
    
    let cgroup_path = format!("{}/{}", CGROUP_ROOT, container_name);
    
    // Create container-specific cgroup directory
    if !std::path::Path::new(&cgroup_path).exists() {
        fs::create_dir_all(&cgroup_path)?;
        fs::set_permissions(&cgroup_path, fs::Permissions::from_mode(0o755))?;
    }
    
    // Set memory limits
    let mem_max_path = format!("{}/memory.max", cgroup_path);
    let mem_swap_path = format!("{}/memory.swap.max", cgroup_path);
    
    match fs::write(&mem_max_path, MEMORY_MAX.to_string()) {
        Ok(_) => {
            eprintln!("✓ Memory limit set: {} bytes ({} MB)", MEMORY_MAX, MEMORY_MAX_MB);
        }
        Err(e) => {
            eprintln!("Warning: Failed to set memory.max limit: {}", e);
        }
    }
    
    match fs::write(&mem_swap_path, MEMORY_SWAP_MAX.to_string()) {
        Ok(_) => {
            eprintln!("✓ Swap limit set: {} bytes ({} MB)", MEMORY_SWAP_MAX, MEMORY_SWAP_MAX_MB);
        }
        Err(e) => {
            eprintln!("Warning: Failed to set memory.swap.max limit: {}", e);
        }
    }
    
    // Set CPU limits
    let cpu_max_path = format!("{}/cpu.max", cgroup_path);
    let cpu_limit = format!("{} {}", CPU_QUOTA_US, CPU_PERIOD_US);
    
    match fs::write(&cpu_max_path, &cpu_limit) {
        Ok(_) => {
            eprintln!("✓ CPU limit set: {} cores max", CPU_QUOTA_US as f64 / CPU_PERIOD_US as f64);
        }
        Err(e) => {
            eprintln!("Warning: Failed to set cpu.max limit: {}", e);
        }
    }
    
    Ok(cgroup_path)
}

pub fn add_process_to_cgroup(cgroup_path: &str, pid: i32) -> io::Result<()> {
    let cgroup_procs = format!("{}/cgroup.procs", cgroup_path);
    fs::write(&cgroup_procs, pid.to_string())?;
    Ok(())
}

pub fn cleanup_cgroup(container_name: &str) -> io::Result<()> {
    let cgroup_path = format!("{}/{}", CGROUP_ROOT, container_name);
    if std::path::Path::new(&cgroup_path).exists() {
        fs::remove_dir(&cgroup_path)?;
    }
    Ok(())
}

/// Get memory usage statistics for a container
pub fn get_memory_stats(container_name: &str) -> io::Result<MemoryStats> {
    let cgroup_path = format!("{}/{}", CGROUP_ROOT, container_name);
    
    let current = fs::read_to_string(format!("{}/memory.current", cgroup_path))?
        .trim()
        .parse::<u64>()
        .unwrap_or(0);
    
    let max = fs::read_to_string(format!("{}/memory.max", cgroup_path))?
        .trim()
        .parse::<u64>()
        .unwrap_or(0);
    
    Ok(MemoryStats {
        current_bytes: current,
        max_bytes: max,
    })
}

pub fn get_memory_from_proc(pid: i32) -> io::Result<u64> {
    let status_path = format!("/proc/{}/status", pid);
    let status_content = fs::read_to_string(status_path)?;
    
    for line in status_content.lines() {
        if line.starts_with("VmRSS:") {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 2 {
                if let Ok(kb) = parts[1].parse::<u64>() {
                    return Ok(kb * 1024); // Convert KB to bytes
                }
            }
        }
    }
    
    Ok(0)
}

/// Get CPU usage percentage from /proc for a specific PID
pub fn get_cpu_from_proc(pid: i32) -> io::Result<f64> {
    let stat_path = format!("/proc/{}/stat", pid);
    let stat_content = fs::read_to_string(stat_path)?;
    
    // Parse /proc/[pid]/stat to get utime + stime (CPU time in clock ticks)
    let parts: Vec<&str> = stat_content.split_whitespace().collect();
    if parts.len() >= 22 {
        // utime is at index 13, stime at index 14, starttime at index 21 (0-indexed)
        let utime = parts[13].parse::<u64>().unwrap_or(0);
        let stime = parts[14].parse::<u64>().unwrap_or(0);
        let starttime = parts[21].parse::<u64>().unwrap_or(0);
        let total_time = utime + stime;
        
        // Get system uptime and calculate process uptime
        let uptime_content = fs::read_to_string("/proc/uptime")?;
        let uptime_parts: Vec<&str> = uptime_content.split_whitespace().collect();
        if let Some(uptime_str) = uptime_parts.first() {
            if let Ok(system_uptime) = uptime_str.parse::<f64>() {
                // Convert everything to seconds (100 ticks/sec)
                let process_start_secs = starttime as f64 / 100.0;
                let process_uptime = system_uptime - process_start_secs;
                
                if process_uptime > 0.0 {
                    let cpu_secs = total_time as f64 / 100.0;
                    // CPU percent = (cpu_time / process_uptime) * 100
                    let cpu_percent = (cpu_secs / process_uptime) * 100.0;
                    return Ok(cpu_percent.min(400.0)); // Cap at 400% (4 cores)
                }
            }
        }
    }
    
    Ok(0.0)
}

#[derive(Debug)]
pub struct MemoryStats {
    pub current_bytes: u64,
    pub max_bytes: u64,
}

impl MemoryStats {
    pub fn current_mb(&self) -> f64 {
        self.current_bytes as f64 / (1024.0 * 1024.0)
    }
    
    pub fn max_mb(&self) -> f64 {
        self.max_bytes as f64 / (1024.0 * 1024.0)
    }
}

File: internal/core/cgroup/cgroup.go
package cgroup

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/Voyrox/Qube/internal/config"
	"github.com/fatih/color"
)

type cpuCache struct {
	totalTime uint64
	timestamp time.Time
}

var cpuCacheMap = make(map[int]*cpuCache)

const (
	MemoryMax     = config.MemoryMaxMB * 1024 * 1024
	MemorySwapMax = config.MemorySwapMaxMB * 1024 * 1024
)

type MemoryStats struct {
	CurrentBytes uint64
}

func (m *MemoryStats) CurrentMB() float64 {
	return float64(m.CurrentBytes) / (1024.0 * 1024.0)
}

func InitCgroupRoot() error {
	if _, err := os.Stat(config.CgroupRoot); os.IsNotExist(err) {
		if err := os.MkdirAll(config.CgroupRoot, 0755); err != nil {
			return err
		}
	}

	subtreeControl := filepath.Join(config.CgroupRoot, "cgroup.subtree_control")
	if err := ioutil.WriteFile(subtreeControl, []byte("+memory +cpu"), 0644); err != nil {
		fmt.Printf("Warning: Failed to enable cgroup controllers: %v\n", err)
	} else {
		color.Green("✓ Cgroup controllers enabled: memory, cpu")
	}

	return nil
}

func SetupCgroupForContainer(containerName string) (string, error) {
	if err := InitCgroupRoot(); err != nil {
		return "", err
	}

	cgroupPath := filepath.Join(config.CgroupRoot, containerName)

	if _, err := os.Stat(cgroupPath); os.IsNotExist(err) {
		if err := os.MkdirAll(cgroupPath, 0755); err != nil {
			return "", err
		}
	}

	memMaxPath := filepath.Join(cgroupPath, "memory.max")
	if err := ioutil.WriteFile(memMaxPath, []byte(fmt.Sprintf("%d", MemoryMax)), 0644); err != nil {
		fmt.Printf("Warning: Failed to set memory.max limit: %v\n", err)
	} else {
		color.Green("✓ Memory limit set: %d bytes (%d MB)", MemoryMax, config.MemoryMaxMB)
	}

	memSwapPath := filepath.Join(cgroupPath, "memory.swap.max")
	if err := ioutil.WriteFile(memSwapPath, []byte(fmt.Sprintf("%d", MemorySwapMax)), 0644); err != nil {
		fmt.Printf("Warning: Failed to set memory.swap.max limit: %v\n", err)
	} else {
		color.Green("✓ Swap limit set: %d bytes (%d MB)", MemorySwapMax, config.MemorySwapMaxMB)
	}

	cpuMaxPath := filepath.Join(cgroupPath, "cpu.max")
	cpuLimit := fmt.Sprintf("%d %d", config.CPUQuotaUS, config.CPUPeriodUS)
	if err := ioutil.WriteFile(cpuMaxPath, []byte(cpuLimit), 0644); err != nil {
		fmt.Printf("Warning: Failed to set cpu.max limit: %v\n", err)
	} else {
		color.Green("✓ CPU limit set: %d/%d", config.CPUQuotaUS, config.CPUPeriodUS)
	}

	return cgroupPath, nil
}

func AddProcessToCgroup(cgroupPath string, pid int) error {
	procsPath := filepath.Join(cgroupPath, "cgroup.procs")
	return ioutil.WriteFile(procsPath, []byte(fmt.Sprintf("%d", pid)), 0644)
}

func GetMemoryStats(containerName string) (*MemoryStats, error) {
	cgroupPath := filepath.Join(config.CgroupRoot, containerName)
	memCurrentPath := filepath.Join(cgroupPath, "memory.current")

	content, err := ioutil.ReadFile(memCurrentPath)
	if err != nil {
		return nil, err
	}

	currentBytes, err := strconv.ParseUint(strings.TrimSpace(string(content)), 10, 64)
	if err != nil {
		return nil, err
	}

	return &MemoryStats{CurrentBytes: currentBytes}, nil
}

func GetMemoryFromProc(pid int) (uint64, error) {
	statusPath := fmt.Sprintf("/proc/%d/status", pid)
	content, err := ioutil.ReadFile(statusPath)
	if err != nil {
		return 0, err
	}

	lines := strings.Split(string(content), "\n")
	for _, line := range lines {
		if strings.HasPrefix(line, "VmRSS:") {
			fields := strings.Fields(line)
			if len(fields) >= 2 {
				rss, err := strconv.ParseUint(fields[1], 10, 64)
				if err == nil {
					return rss * 1024, nil
				}
			}
		}
	}

	return 0, fmt.Errorf("VmRSS not found")
}

func GetCPUFromProc(pid int) (float64, error) {
	statPath := fmt.Sprintf("/proc/%d/stat", pid)
	content, err := ioutil.ReadFile(statPath)
	if err != nil {
		delete(cpuCacheMap, pid)
		return 0, err
	}

	fields := strings.Fields(string(content))
	if len(fields) < 15 {
		return 0, fmt.Errorf("invalid stat format")
	}

	utime, _ := strconv.ParseUint(fields[13], 10, 64)
	stime, _ := strconv.ParseUint(fields[14], 10, 64)
	totalTime := utime + stime

	now := time.Now()

	cached, exists := cpuCacheMap[pid]
	if !exists {
		cpuCacheMap[pid] = &cpuCache{
			totalTime: totalTime,
			timestamp: now,
		}
		return 0.0, nil
	}

	timeDelta := now.Sub(cached.timestamp).Seconds()
	if timeDelta < 0.1 {
		return 0.0, nil
	}

	cpuDelta := totalTime - cached.totalTime
	cpuPercent := (float64(cpuDelta) / 100.0 / timeDelta) * 100.0

	cpuCacheMap[pid] = &cpuCache{
		totalTime: totalTime,
		timestamp: now,
	}

	return cpuPercent, nil
}

func RemoveCgroup(containerName string) error {
	cgroupPath := filepath.Join(config.CgroupRoot, containerName)
	return os.RemoveAll(cgroupPath)
}
